client/trino-client/src/main/java/io/trino/client/ResultRowsDecoder.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client;
+
+import com.google.common.collect.Iterators;
+import io.trino.client.spooling.DataAttributes;
+import io.trino.client.spooling.EncodedQueryData;
+import io.trino.client.spooling.InlineSegment;
+import io.trino.client.spooling.Segment;
+import io.trino.client.spooling.SegmentLoader;
+import io.trino.client.spooling.SpooledSegment;
+import io.trino.client.spooling.encoding.QueryDataDecoders;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.client.FixJsonDataUtils.fixData;
+import static io.trino.client.ResultRows.NULL_ROWS;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Class responsible for decoding any QueryData type.
+ */
+public class ResultRowsDecoder
+        implements AutoCloseable
+{
+    private final SegmentLoader loader;
+    private final AtomicReference<QueryDataDecoder.Factory> decoderFactory = new AtomicReference<>();
```

Comment:
Not strictly related to this PR.
But I do not see why you need those `Atomic*` here. 
What thread safety does it provide, where do we expect this can be used from multiple thread where explicit memory barrier does not exist between`ResultRowsDecoder` initialization and usage.